import os
import pandas as pd
import glob
configfile: '../shared/config.yaml'
include: "../shared/Snakefile"

#rule all:
#    input:
#        "/scr/dings/PEVO/NEW_downloads_3/TCGA-pipelines/TCGA-CESC_TCGA-HNSC/metilene/metilene_input_table/carboplatn,paclitaxel_cisplatin/male/cutoff_0/threshold_0/summary_for_metilene.tsv"
#breakpoint()
def get_data_files(wildcards):
    OUTPUT_PATH = wildcards[0]
    project = wildcards[1].split('_')
    DF_temp = pd.read_table(os.path.join(OUTPUT_PATH, 'metadata', config['manifest_file'][0]))
    DF_temp.set_index('project_id').loc[project, 'filename'].str.contains('HumanMethylation450')
    DF_temp = DF_temp[DF_temp['filename'].str.contains(
        'HumanMethylation450')].set_index(
                'project_id').loc[project, 'filename']
    DF_temp = DF_temp.to_frame()
    DF_temp.reset_index(inplace=True)
    DF_temp['filepath'] = OUTPUT_PATH + '/' + DF_temp['project_id'] + '/metilene/'+ 'data_files' + '/' + DF_temp['filename']
    data_path_list = DF_temp['filepath'].to_list()
    data_path_list = [ i + '.gz' for i in data_path_list]
    return data_path_list

rule create_summary_table:
    """
    according to the flag composition the (projects, drugs) and the metadata
    given (dead/alive, female/male) create metilene input tables
    # every data file for the belonging project must be downloaded, since not
    # just the selection of drugs is invoked, but also the complement
    """
    input:
        meta_file = "{output_path}/{project}/metilene/merged_meta_files/{cutoff}/meta_info_druglist_merged_drugs_combined.tsv",
        manifest_file = expand("{{output_path}}/metadata/{manifest}", manifest=config['manifest_file']),
        data_files = get_data_files,
    output:
        out_file = "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene.tsv",
        out_file_complement = "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene_complement.tsv",
    conda:
        "../shared/envs/pandas_natsort.yaml"
    threads:  # although no threads are applied, limit the process since the table creations needs a lot of ram
        5
    log:
        "{output_path}/{project}/metilene/logs/create_summary_table/{drug_combi}_{gender}_{cutoff}.log",
    script:
        "scripts/create_summary_table.py"

rule create_metilene_out:
    """
    # we do not use the conda metilene but use our provided bin and script:
    # the commands could be:
    metilene -m 3 -M 1000 -d 0.03 -a alive -b dead summary_for_metilene.tsv | sort -V -k1,1 -k2,2n > metilene_out_sorted.tsv
    perl metilene_output.pl -q metilene_out_sorted.tsv -a 'alive' -b 'dead' -c 3 -d 0.03
    # for both metilene and the filter script:
    -d	minimum mean methylation difference (>=) (default:0.1)
    # metilene specific:
    -M, --maxdist	Integer	300	The allowed nt distance between two CpGs within a DMR
    -m, --mincpgs	Integer	10	The minimum # of CpGs in a DMR
    # filter specific
    -c	minimum (>=) CpGs (default:10)
    we apply first alive, then dead, 
    a negative mean methylation difference is going to be alive up, a positive means alive down

    """
    input:
        "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene.tsv",
    output:
        "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_out_sorted.tsv", 
        "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_qval.0.05.out", 
        "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_qval.0.05.pdf", 
        "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_qval.0.05.bedgraph", 
    conda:
        "envs/metilene.yaml"
    threads:
        10
    log:
        "{output_path}/{project}/metilene/logs/create_metilene_out/{drug_combi}_{gender}_{cutoff}.log",
    conda:
        "envs/pandas2.yaml"
    script:
        "scripts/create_metilene_out.py"

rule metilene_intersect_tables:
    """
    intersecting the qval.0.05.out tables (holding DMRs) with the beta value files with which the are created with
    This intersect file is the base for many following analyses
    """
    input:
        qval_out = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_qval.0.05.out", 
        #meta_info = "{output_path}/{project}/metilene/merged_meta_files/meta_info_druglist_merged_drugs_combined.tsv",
        # is this actually needed?
        metilene_input = "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene.tsv",
    output:
        out_file = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect.tsv"
    conda:
        "envs/pandas_bedtools.yaml"
    log:
        "{output_path}/{project}/metilene/logs/metilene_intersect_tables/{drug_combi}_{gender}_{cutoff}.log",
    script:
        "scripts/bed_intersect_metilene.py"

rule metilene_plot_DMR_regions:
    """
    based on the DMRs within the intersect_tables, those DMRs are plotted both,
    as box and lineplots:
    """
    input:
        metilene_intersect = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect.tsv"
    output:
        pdf_boxplot_out = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect_boxplot_beta_value_{DMR}.pdf",
        pdf_lineplot_out = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect_lineplot_median_beta_value_{DMR}.pdf" 
    conda:
        "envs/seaborn_plot.yaml"
    log:
        "{output_path}/{project}/metilene/logs/metilene_plot_DMR_regions/{drug_combi}_{gender}_{cutoff}_{DMR}.log",
    script:
        "scripts/plot_DMR_regions.py"

rule create_metilene_lifeline_plots:
    """
    in the validation and evaluation steps, those plots are called "base_plot"
    # here the ENSGs are also directly added based on chr and start position and the annot file
    # this is important for the validation steps
    # important to keep the increasing or decreasing info out of the metilene_intersect table
    the information of mean methylation difference must also be tracked -> if its negative, the filename is INCREASING, pos -> DECREASING
    """
    input:
        metilene_intersect = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect.tsv",
        meta_table = "{output_path}/{project}/metilene/merged_meta_files/{cutoff}/meta_info_druglist_merged_drugs_combined.tsv",
        annot_file = expand("{{output_path}}/metadata_processed/{gtf}_genes_transcripts.gz", gtf=config['gtf']),
        annot_file_2 = "../shared/resources/annot_from_betafile.tsv.gz",
        script_file = "../tcga_metilene/scripts/create_lifeline_plots.py"
    output:
        lifeline_out_pdf = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}.pdf",
        lifeline_out_tsv = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}.tsv",
    wildcard_constraints:
        DMR=".*_.*_\d*"  # to avoid that output files with pattern #metilene_intersect_lifeline_plot_chr19_57351041_57351345_UP_val.pdf match this rule
                        # -> limit the wildcards to chrxy_start_end pattern, otherwise DMR could be interpreted as chr19_57351041_57351345_UP_val
    log:
        "{output_path}/{project}/metilene/logs/create_lifeline_plots/{drug_combi}_{gender}_{cutoff}_{threshold}_{DMR}.log"
    conda:
        "envs/lifelines.yaml"
    script:
        "scripts/create_lifeline_plots.py"

rule DMR_plot_validation:
    """
    in the validation and evaluation steps, those plots are called "UP_validation" and "DOWN_validation"
    # here the ENSGs are also directly added based on chr and start position and the annot file
    # this is important for the validation steps
    # there are regions which are not annotated within the gtf, but in the datafiles:
    bedtools intersect -b <(echo "chr11\t115505380\t115505381") -a  /scr/palinca/gabor/TCGA-pipeline/metadata/gencode.v36.annotation.gtf.gz
    bedtools intersect -b <(echo "chr11\t115505380\t115505381") -a <(awk 'NR>1{print $3"\t"$4"\t"$5"\t"$0}' /scr/palinca/gabor/TCGA-pipeline/TCGA-CESC/metilene/data_files/jhu-usc.edu_CESC.HumanMethylation450.1.lvl-3.TCGA-C5-A1BE-01B-11D-A13Z-05.gdc_hg38.txt | grep -v "*")
    chr11   115505380       115505381       cg01948062      0.0948425263590688      chr11   115505380       115505381       CADM1;CADM1;CADM1;CADM1;CADM1;CADM1;CADM1;CADM1;CADM1;CADM1;CADM1       protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding       ENST00000331581.9;ENST00000452722.6;ENST00000536727.4;ENST00000536781.1;ENST00000537058.4;ENST00000537140.4;ENST00000540951.1;ENST00000541434.4;ENST00000542447.5;ENST00000543249.1;ENST00000545380.4      -814;-964;-964;-964;-964;-856;-964;-985;-856;-422;-990  CGI:chr11:115502856-115505163   S_Shore
    # the ENST and gene symbol can also not be found within the gtf
    grep CADM1 /scr/palinca/gabor/TCGA-pipeline/metadata/gencode.v36.annotation.gtf.gz
    grep ENST00000331581 /scr/palinca/gabor/TCGA-pipeline/metadata/gencode.v36.annotation.gtf.gz

    based on the found DMRs, check those DMRs for the following sets:
                                               |  initial DMR  |      
                                               |---------------|
                                               |               | 
                                               |     ja        |
                                               |   /    \      |
                                    in therapy?|--        KM   |   
                                   /           |   \    /      |
                               hoch  50%       |    nein       |
                              /      |         |               |
    - KP including every case:  ---median      |               |
                              \      |         |               |
                               tief  50%       |     ja        |
                                   \           |   /    \      |
                                    in therapy?|--        KM   |   
                                               |   \    /      |   
                                               |    nein       |

    - the initial DMR KM plot groups beta value UP and DOWN for treated cases
    - the two validation KM plots groups treated vs untreated for all cases which are grouped in UP and DOWN by median

    """
    input:
        meta_table = "{output_path}/{project}/metilene/merged_meta_files/{cutoff}/meta_info_druglist_merged_drugs_combined.tsv",
        # this table holds the desired Start value:
        start_tsv = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}.tsv",
        # every betavalue available is hold by those two tables, filter on threshold withing validation script:
        summary = "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene.tsv",
        summary_complement = "{output_path}/{project}/metilene/metilene_input_table/{drug_combi}/{gender}/{cutoff}/summary_for_metilene_complement.tsv",
        annot_file = expand("{{output_path}}/metadata_processed/{gtf}_genes_transcripts.gz", gtf=config['gtf']),
        annot_file_2 = "../shared/resources/annot_from_betafile.tsv.gz",
        script_file = "../tcga_metilene/scripts/create_lifeline_plots_validation.py"
    output:
        UP_val_plot = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}_UP_val.pdf",
        UP_val_tsv = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}_UP_val.tsv",
        DOWN_val_plot = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}_DOWN_val.pdf",
        DOWN_val_tsv = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{DMR}_DOWN_val.tsv",
    conda:
        "envs/lifelines.yaml"
    log:
        "{output_path}/{project}/metilene/logs/DMR_plot_validation/{drug_combi}_{gender}_{cutoff}_{threshold}_{DMR}.log"
    script:
        "scripts/create_lifeline_plots_validation.py"

#merge  the lifeline pdfs together:

def get_plots_to_aggregate(wildcards):
    # before this input fct is required, the metilene_intersect tables must be
    # created and with that, the DMRs on which we rely are known, 'tell' which
    # DMR plots we expect: - it's not sufficient to glob the wildcards, since
    # it's not assured that those files are already created!
    path = os.path.join(wildcards[0], wildcards[1], 'metilene/metilene_output', wildcards[2], wildcards[3], wildcards[4], 'metilene_intersect.tsv')
    try:
        DMRs = pd.read_table(path,
                header=[0,1,2,3,4], index_col=[0,1,2,3],
                na_values='.').reset_index(
                        'region')['region'].value_counts().index.tolist()
    except Exception as e:
        "in case the intersect table is empty, like in CESC for male, "
        "return an empty list"
        print(e)
        print(f'{path} containts no DMRs, no merge plot can be created')
        return []
    beta_plots = sorted([path.replace('.tsv', '_boxplot_beta_value_' + dmr + '.pdf') for dmr in DMRs])
    line_plots = sorted([path.replace('.tsv', '_lineplot_median_beta_value_' + dmr + '.pdf') for dmr in DMRs])
    # wildcards[5] is the threshold which is invokened with the lifeline plot
    lifeline_template = os.path.join(wildcards[0], wildcards[1], 'metilene/metilene_output', wildcards[2], wildcards[3], wildcards[4], wildcards[5], 'metilene_intersect_lifeline_plot_')
    lifeline_plots = sorted([(lifeline_template + dmr + '.pdf') for dmr in DMRs])
    lifeline_plots_UP = sorted([(lifeline_template + dmr + '_UP_val.pdf') for dmr in DMRs])
    lifeline_plots_DOWN = sorted([(lifeline_template + dmr +'_DOWN_val.pdf') for dmr in DMRs])
    final_list = []
    for i,j,x,y,z in zip(beta_plots, line_plots, lifeline_plots, lifeline_plots_UP, lifeline_plots_DOWN):
        final_list.append(i)
        final_list.append(j)
        final_list.append(x)
        final_list.append(y)
        final_list.append(z)
    #print(f'merging {final_list}')
    return final_list
    #glob.glob(f'{wildcards.output_path}/{wildcards.project}/metilene/metilene_output/{wildcards.drug_combi}/{wildcards.gender}/{wildcards.cutoff_and_or_threshold}/metilene_intersect_{wildcards.plot_type}_*.pdf')

rule merge_DMR_plots:
    '''
    # we can provide 
    merging every plot which belongs to one DMR
    # the shell: directive would work, but the sorting is not preferable, so
    # using python to group every DMR kind together, also the complement plots:
    # here, we do not invoke pdf's to the merge which are to small (empty):
    qpdf --empty --pages $(find . -maxdepth 1 -type f -size +1k -iname '*intersect*pdf') -- all_plots_merged.pdf
    [Errno 7] Argument list too long: '/bin/sh'
    # avoid a too long argument list with invokening python pdf funcitonality,
    # do not use bash tools like qpdf or pdfcombine
    # if the pdf list extends the length of 100, merge them beforehand, and
    # than merge those merged pdfs, otherwise to many open files error occurs
    '''
    input:
        intersect_file = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect.tsv",
        plots_to_aggregate = get_plots_to_aggregate, # with this input function, it is secured that the plots are created before merging
    output:
        All_plots_merged = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/All_plots_merged.pdf",
    conda:
        "envs/pypdf.yaml"
    log:
        "{output_path}/{project}/metilene/logs/merge_DMR_plots/{drug_combi}_{gender}_{cutoff}_{threshold}.log"
    script:
        "scripts/merge_DMR_region_plots.py"
    #shell:
        #"qpdf --empty --pages {input} -- {output}"
        #qpdf --empty --pages Linux.pdf ShellIntro.pdf -- new.pdf
        #find . -maxdepth 1 -type f -size +1k -iname '*intersect*pdf'

def get_threshold_list():
    with open("../tcga_metilene/dynamic/thresholds.txt") as f:
        return [line.strip() for line in f]
threshold_list = get_threshold_list()
# with every call of this script, the threshold_str is created, based on the
# cli options set
threshold_str = "_".join(threshold_list)

def get_lifeline_all(wildcards):
    """
    those single tables shall be returned:
        base_plot:
        metilene_intersect_lifeline_plot_chr2_176150267_176153069.tsv         
        metilene_lifeline_tsv = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{chr}_{start}_{end}.tsv",

        validation_plots:
        metilene_intersect_lifeline_plot_chr2_176150267_176153069_DOWN_val.tsv
        metilene_intersect_lifeline_plot_chr2_176150267_176153069_UP_val.tsv  
        metilene_lifeline_tsv_UP = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{chr}_{start}_{end}_UP.tsv",
        metilene_lifeline_tsv_DOWN = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/{threshold}/metilene_intersect_lifeline_plot_{chr}_{start}_{end}_DOWN.tsv",
    # the thresholds are merged, explicitly naming them via wildcards is not possible, glob them
    """
    output_path = wildcards.output_path
    project = wildcards.project
    drug_combi = wildcards.drug_combi
    gender = wildcards.gender
    cutoff = wildcards.cutoff
    # the glob is still necessary but just include applied thresholds within the dynamic/thresholds.txt file.
    search_result = []
    for thresh in threshold_list:
        search_result = search_result + glob.glob(os.path.join(output_path, project, "metilene/metilene_output", drug_combi, gender, cutoff, f'{thresh}/metilene_intersect_lifeline*.tsv' ))
    return search_result

rule aggregate_lifelines_all:
    """
    metilene out informations like mean methylation differences, q-value and #DMRs are added aswell
    """
    input:
        # not using a input fct, we rather expand here the thresholds which should be aggregated explicitly, with that its still possible to get different aggregated results for different threshold combinations through the cli:
        # so the threshold inclusion depends on the interface threshold input in each single run
        lifeline_all = get_lifeline_all, # first lifeline plots with therapy set invoked and validation plot for UP and for DOWN regulated expressions
        metilene_out = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_qval.0.05.out", 
        # the results are not needed within the aggregate_lifelines_all.py script, but shall be a precondition to run this rule
        #metilene_results = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/metilene_log2f{in_de}CREASE_result.tsv", in_de=['IN', 'DE']),
        #metilene_results = "{output_path}/{project}/metilene/metilene_output/{drug_combi}/{gender}/{cutoff}/metilene_intersect.tsv"
        script_file = "../tcga_metilene/scripts/aggregate_lifelines_all.py"
    output:
        metilene_lifeline_aggregated = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_lifelines_aggregated.tsv.gz", threshold_str=threshold_str),
        #metilene_lifeline_aggregated = "/scr/metilene_heatmap_log2fDECREASE_raw_counts_aggregated.tsv",
    conda:
        "envs/pandas2.yaml"
    threads:
        1
    log:
        "{output_path}/{project}/metilene/logs/aggregate_lifelines_all/{drug_combi}_{gender}_{cutoff}.log",
    script:
        "scripts/aggregate_lifelines_all.py"

rule evaluate_lifelines_all:
    input:
        metilene_lifeline_aggregated = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_lifelines_aggregated.tsv.gz", threshold_str=threshold_str),
    output:
        metilene_lifeline_eval = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_lifelines_evaluated.tsv.gz", threshold_str=threshold_str),
        metilene_lifeline_eval_pdfs = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_lifelines_evaluated.pdf", threshold_str=threshold_str),
        #metilene_lifeline_aggregated = "/scr/metilene_heatmap_log2fDECREASE_raw_counts_aggregated.tsv",
    conda:
        "envs/pandas_pypdf.yaml"
    threads:
        1
    log:
        "{output_path}/{project}/metilene/logs/evaluate_lifelines_all/{drug_combi}_{gender}_{cutoff}.log",
    script:
        "scripts/evaluate_lifelines_all.py"

#TODO this can be merged in the shared Snakefile, the only difference is
#depending on the pipeline, what to use as xticks, in metilene, tist the chr -
#start and in deseq the ensg 
rule plot_diffs:
    """
    for every found ensg, plot the pval or the life mean diff, also together,
    if applied, with the different thresholds invoked
    """
    input:
        metilene_lifeline_aggregated = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_lifelines_aggregated.tsv.gz", threshold_str=threshold_str),
        script_file = "../tcga_metilene/scripts/plot_thresh_diff.py"
    output:
        plot_diffs = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_plot_diffs.pdf", threshold_str=threshold_str),
        plot_diffs_table = expand("{{output_path}}/{{project}}/metilene/metilene_output/{{drug_combi}}/{{gender}}/{{cutoff}}/{threshold_str}/metilene_plot_diffs.tsv.gz", threshold_str=threshold_str),
    conda:
        "envs/seab_matplot_plotl.yaml"
    threads:
        1
    log:
        "{output_path}/{project}/metilene/logs/plot_diffs/{drug_combi}_{gender}_{cutoff}.log",
    script:
        "scripts/plot_thresh_diff.py"
